File: HackerEarth/minimum nodes.cpp
#include<bits/stdc++.h>
using namespace std;
const int mx = 1e5+10;

vector<int> gr[mx];
long long val[mx];
int u,v,q,n,x,i,ans1;

void dfs(int node,int cnt,int par,long long sum1,long long k) {
    sum1 += val[node];
    cnt++;
    if( sum1 >= k ){
        ans1 = min(ans1,cnt);
    }
    for(int i=0; i<gr[node].size(); i++) {
        int nxt=gr[node][i];
        if(nxt == par) {
            continue;
        }
        dfs(nxt,cnt,node,sum1,k);
    }
}

int main() {
    long long k;
    scanf("%d %d",&n, &q);
    for(int i=0; i<n; i++) {
        scanf("%d",&val[i]);
    }
    for(int i=1; i<=n-1; i++) {
        scanf("%d %d",&u, &v);
        gr[u].push_back(v);
        gr[v].push_back(u);
    }
    for(int i=0; i<q; i++) {
        scanf("%d %lld",&x, &k);
        ans1 = INT_MAX;
        dfs(x,0,0,0,k);
        if(ans1 == INT_MAX) {
            ans1=-1;
        }
        printf("%d\n",ans1);
    }

    return 0;
}

File: HackerEarth/minimum_nodes.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;

public class minimum_nodes {
    static int max_node = 100005;
    static ArrayList<Integer> graph[];
    static long[] val = new long[max_node];
    static int ans = Integer.MAX_VALUE;

    static void dfs(int node,int cnt,int par,long sum,long k) {
        sum += val[node];
        cnt += 1;
        if(sum >= k) {
            ans = Integer.min(ans, cnt);
        }
        for(int it: graph[node]) {
            if(it == par) {
                continue;
            }
            dfs(it,cnt,node,sum,k);
        }
    }

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] str = br.readLine().split(" ");
        int n = Integer.parseInt(str[0]);
        int q = Integer.parseInt(str[1]);

        // initialize the graph with nodes
        graph = new ArrayList[n+1];

        String[] str1 = br.readLine().split(" ");
        for(int i=0; i < str1.length; i++) {
            val[i+1] = Long.parseLong(str1[i]);
            // initialize each node with an Arraylist<>
            graph[i+1] = new ArrayList<Integer>();
        }

        for(int i=1; i<=n-1; i++) {
            str = br.readLine().split(" ");
            int u = Integer.parseInt(str[0]);
            int v = Integer.parseInt(str[1]);
            graph[u].add(v);
            graph[v].add(u);
        }

        for(int i=1; i<=q; i++) {
            str = br.readLine().split(" ");
            int x = Integer.parseInt(str[0]);
            long k = Long.parseLong(str[1]);
            ans = Integer.MAX_VALUE;
            dfs(x,0,0,0,k);
            if(ans == Integer.MAX_VALUE) {
                ans = -1;
            }
            System.out.println(ans);
        }

    }
}
